File: src/precalculus_function_composition_hard.py
"""
precalculus - function_composition (hard)
Generated: 2026-02-22T04:38:27.619215
"""

from problem_utils import *

def generate():
    problem_type = choice(['nested_evaluate', 'nested_solve', 'decompose', 'inverse_composition', 'multi_composition'])
    
    if problem_type == 'nested_evaluate':
        # Given f(x) and g(x), find f(g(a)) for specific value a
        # ELO: 1600-1700
        a_coeff = nonzero(-3, 3)
        b_const = nonzero(-5, 5)
        c_coeff = nonzero(-4, 4)
        d_const = nonzero(-6, 6)
        
        # f(x) = a*x + b, g(x) = c*x^2 + d
        f_expr = a_coeff * x + b_const
        g_expr = c_coeff * x**2 + d_const
        
        eval_point = nonzero(-3, 3)
        
        # Calculate g(eval_point)
        g_val = g_expr.subs(x, eval_point)
        # Calculate f(g(eval_point))
        ans = f_expr.subs(x, g_val)
        
        return problem(
            question=f"Let $f(x) = {latex(f_expr)}$ and $g(x) = {latex(g_expr)}$. Find $f(g({eval_point}))$.",
            answer=ans,
            difficulty=(1600, 1700),
            topic="precalculus/function_composition",
            solution=steps(
                f"First evaluate $g({eval_point})$:",
                f"$g({eval_point}) = {latex(c_coeff)}({eval_point})^2 + {latex(d_const)} = {latex(g_val)}$",
                f"Now evaluate $f({latex(g_val)})$:",
                f"$f({latex(g_val)}) = {latex(a_coeff)}({latex(g_val)}) + {latex(b_const)} = {latex(ans)}$"
            ),
        )
    
    elif problem_type == 'nested_solve':
        # Given f(x) and g(x), solve f(g(x)) = k for x
        # ELO: 1700-1800
        a_coeff = nonzero(-2, 2)
        b_const = nonzero(-4, 4)
        c_coeff = choice([1, 2])
        d_const = nonzero(-3, 3)
        
        # f(x) = a*x + b, g(x) = c*x + d
        f_expr = a_coeff * x + b_const
        g_expr = c_coeff * x + d_const
        
        # Compose f(g(x))
        fg_expr = f_expr.subs(x, g_expr)
        fg_simplified = simplify(fg_expr)
        
        # Pick a target value k
        target_k = nonzero(-10, 10)
        
        # Solve f(g(x)) = k
        solutions = solve(Eq(fg_simplified, target_k), x)
        ans = solutions[0]
        
        return problem(
            question=f"Let $f(x) = {latex(f_expr)}$ and $g(x) = {latex(g_expr)}$. Solve $f(g(x)) = {target_k}$ for $x$.",
            answer=ans,
            difficulty=(1700, 1800),
            topic="precalculus/function_composition",
            solution=steps(
                f"First find $f(g(x))$:",
                f"$f(g(x)) = f({latex(g_expr)}) = {latex(a_coeff)}({latex(g_expr)}) + {latex(b_const)}$",
                f"$f(g(x)) = {latex(fg_simplified)}$",
                f"Now solve ${latex(fg_simplified)} = {target_k}$:",
                f"$x = {latex(ans)}$"
            ),
        )
    
    elif problem_type == 'decompose':
        # Given h(x), find f and g such that h(x) = f(g(x))
        # ELO: 1750-1850
        inner_coeff = nonzero(-3, 3)
        inner_const = nonzero(-4, 4)
        outer_power = choice([2, 3])
        
        # h(x) = (inner_coeff*x + inner_const)^outer_power
        inner_expr = inner_coeff * x + inner_const
        h_expr = inner_expr**outer_power
        
        g_ans = inner_expr
        f_ans = x**outer_power
        
        return problem(
            question=f"Find functions $f(x)$ and $g(x)$ such that $f(g(x)) = {latex(expand(h_expr))}$ where $g(x)$ is linear. Express $g(x)$ in simplest form.",
            answer=g_ans,
            difficulty=(1750, 1850),
            topic="precalculus/function_composition",
            solution=steps(
                f"We need to decompose ${latex(expand(h_expr))}$ as a composition.",
                f"Notice that ${latex(expand(h_expr))} = ({latex(inner_expr)})^{outer_power}$",
                f"Let $g(x) = {latex(inner_expr)}$ and $f(x) = x^{outer_power}$",
                f"Then $f(g(x)) = ({latex(inner_expr)})^{outer_power} = {latex(expand(h_expr))}$",
                f"Therefore, $g(x) = {latex(g_ans)}$"
            ),
        )
    
    elif problem_type == 'inverse_composition':
        # Given f(x) and g(x), find f^(-1)(g(a))
        # ELO: 1800-1900
        a_coeff = nonzero(-3, 3)
        b_const = nonzero(-5, 5)
        c_coeff = choice([1, 2, 3])
        d_const = nonzero(-4, 4)
        
        # f(x) = a*x + b (invertible), g(x) = c*x + d
        f_expr = a_coeff * x + b_const
        g_expr = c_coeff * x + d_const
        
        # Find f^(-1)(x): solve y = a*x + b for x
        f_inv = solve(Eq(y, f_expr), x)[0]
        f_inv_expr = f_inv.subs(y, x)
        
        eval_point = nonzero(-2, 2)
        g_val = g_expr.subs(x, eval_point)
        ans = f_inv_expr.subs(x, g_val)
        
        return problem(
            question=f"Let $f(x) = {latex(f_expr)}$ and $g(x) = {latex(g_expr)}$. Find $f^{{-1}}(g({eval_point}))$.",
            answer=ans,
            difficulty=(1800, 1900),
            topic="precalculus/function_composition",
            solution=steps(
                f"First find $f^{{-1}}(x)$ by solving $y = {latex(f_expr)}$ for $x$:",
                f"$x = {latex(f_inv_expr)}$, so $f^{{-1}}(x) = {latex(f_inv_expr)}$",
                f"Next evaluate $g({eval_point})$:",
                f"$g({eval_point}) = {latex(g_val)}$",
                f"Finally, evaluate $f^{{-1}}({latex(g_val)})$:",
                f"$f^{{-1}}({latex(g_val)}) = {latex(ans)}$"
            ),
        )
    
    else:  # multi_composition
        # Given f(x), g(x), h(x), find f(g(h(a)))
        # ELO: 1650-1750
        a_coeff = nonzero(-2, 2)
        b_const = nonzero(-3, 3)
        c_const = nonzero(-4, 4)
        d_coeff = choice([1, 2])
        
        # f(x) = a*x + b, g(x) = x^2 + c, h(x) = d*x
        f_expr = a_coeff * x + b_const
        g_expr = x**2 + c_const
        h_expr = d_coeff * x
        
        eval_point = nonzero(-3, 3)
        
        h_val = h_expr.subs(x, eval_point)
        g_val = g_expr.subs(x, h_val)
        ans = f_expr.subs(x, g_val)
        
        return problem(
            question=f"Let $f(x) = {latex(f_expr)}$, $g(x) = {latex(g_expr)}$, and $h(x) = {latex(h_expr)}$. Find $f(g(h({eval_point})))$.",
            answer=ans,
            difficulty=(1650, 1750),
            topic="precalculus/function_composition",
            solution=steps(
                f"Evaluate from the inside out. First find $h({eval_point})$:",
                f"$h({eval_point}) = {latex(h_val)}$",
                f"Next find $g({latex(h_val)})$:",
                f"$g({latex(h_val)}) = ({latex(h_val)})^2 + {latex(c_const)} = {latex(g_val)}$",
                f"Finally find $f({latex(g_val)})$:",
                f"$f({latex(g_val)}) = {latex(a_coeff)}({latex(g_val)}) + {latex(b_const)} = {latex(ans)}$"
            ),
        )

emit(generate())

File: src/precalculus_function_composition_hard.jl
# precalculus - function_composition (hard)
# Generated: 2026-03-08T20:24:33.625926

include(joinpath(@__DIR__, "..", "julia", "src", "ProblemUtils.jl"))
using .ProblemUtils

@script x y begin
    set_topic!("precalculus/function_composition")
    
    problem_type = rand(1:5)
    
    if problem_type == 1
        # Type 1: Compose two polynomials and find (f∘g)(a)
        f = rand_quadratic(x; a=(-4,4), b=(-8,8), c=(-12,12))
        g = rand_linear(y; a=(-5,5), b=(-8,8))
        
        a_val = randint(-6, 6)
        g_at_a = substitute(g.expr, y => a_val)
        fog_at_a = substitute(f.expr, x => g_at_a)
        
        problem(
            question="Let \$f(x) = $(tex(f.expr))\$ and \$g(y) = $(tex(g.expr))\$. Find \$(f \\circ g)($(a_val))\$.",
            answer=fog_at_a,
            difficulty=(1800, 2000),
            solution=steps(
                sol("First find", "g($(a_val)) = $(tex(g.expr)) = $(g_at_a)", y => a_val),
                sol("Then compute", "f($(g_at_a)) = $(tex(f.expr))", x => g_at_a),
                sol("Simplify to get", fog_at_a)
            ),
            time=120
        )
        
    elseif problem_type == 2
        # Type 2: Find f∘g as a function, then evaluate
        f_coeffs = [nonzero(-3,3), randint(-10,10)]
        g = rand_quadratic(x; a=(-3,3), b=(-7,7), c=(-10,10))
        
        f_expr = f_coeffs[1]*x + f_coeffs[2]
        fog_expr = substitute(f_expr, x => g.expr)
        fog_expanded = expand(fog_expr)
        
        eval_at = randint(-4, 4)
        answer = substitute(fog_expanded, x => eval_at)
        
        problem(
            question="Given \$f(x) = $(tex(f_expr))\$ and \$g(x) = $(tex(g.expr))\$, find \$(f \\circ g)(x)\$ and evaluate it at \$x = $(eval_at)\$.",
            answer=answer,
            difficulty=(1900, 2100),
            solution=steps(
                sol("Compute composition", "(f \\circ g)(x) = f(g(x)) = $(tex(fog_expr))"),
                sol("Expand", fog_expanded),
                sol("Evaluate at x = $(eval_at)", answer)
            ),
            time=180
        )
        
    elseif problem_type == 3
        # Type 3: Solve f(g(x)) = k for x
        g = rand_linear(x; a=(-4,4), b=(-9,9))
        f = rand_linear(y; a=(-5,5), b=(-10,10))
        
        target_x = randint(-8, 8)
        g_val = substitute(g.expr, x => target_x)
        k = substitute(f.expr, y => g_val)
        
        fog_expr = substitute(f.expr, y => g.expr)
        
        problem(
            question="Let \$f(y) = $(tex(f.expr))\$ and \$g(x) = $(tex(g.expr))\$. Solve \$(f \\circ g)(x) = $(k)\$ for \$x\$.",
            answer=target_x,
            difficulty=(1850, 2050),
            solution=steps(
                sol("Set up equation", "(f \\circ g)(x) = $(tex(fog_expr)) = $(k)"),
                sol("Solve linear equation", "$(tex(g.a*f.a))*x + $(f.a*g.b + f.b) = $(k)"),
                sol("Solution", target_x)
            ),
            time=150
        )
        
    elseif problem_type == 4
        # Type 4: Find g such that f(g(x)) = h(x)
        f = rand_linear(y; a=(-6,6), b=(-12,12))
        target_g = rand_linear(x; a=(-5,5), b=(-10,10))
        h_expr = substitute(f.expr, y => target_g.expr)
        h_expanded = expand(h_expr)
        
        problem(
            question="Let \$f(y) = $(tex(f.expr))\$ and \$h(x) = $(tex(h_expanded))\$. Find the linear function \$g(x) = ax + b\$ such that \$(f \\circ g)(x) = h(x)\$. Enter your answer as the coefficient \$a\$.",
            answer=target_g.a,
            difficulty=(2000, 2200),
            solution=steps(
                sol("We need", "f(g(x)) = $(tex(f.a))(ax + b) + $(tex(f.b)) = $(tex(h_expanded))"),
                sol("Expand", "$(tex(f.a))*a*x + $(tex(f.a))*b + $(tex(f.b)) = $(tex(h_expanded))"),
                sol("Match coefficients", "$(tex(f.a))*a = $(tex(h_expanded.coeff(x))), so a = $(target_g.a)")
            ),
            time=240
        )
        
    else
        # Type 5: Triple composition f(g(h(x)))
        f = rand_linear(x; a=(-4,4), b=(-8,8))
        g = rand_linear(y; a=(-4,4), b=(-8,8))
        h = rand_linear(x; a=(-4,4), b=(-8,8))
        
        eval_at = randint(-5, 5)
        
        h_val = substitute(h.expr, x => eval_at)
        g_val = substitute(g.expr, y => h_val)
        f_val = substitute(f.expr, x => g_val)
        
        problem(
            question="Let \$f(x) = $(tex(f.expr))\$, \$g(y) = $(tex(g.expr))\$, and \$h(x) = $(tex(h.expr))\$. Find \$(f \\circ g \\circ h)($(eval_at))\$.",
            answer=f_val,
            difficulty=(2100, 2300),
            solution=steps(
                sol("First", "h($(eval_at)) = $(tex(h.expr)) = $(h_val)", x => eval_at),
                sol("Then", "g($(h_val)) = $(tex(g.expr)) = $(g_val)", y => h_val),
                sol("Finally", "f($(g_val)) = $(tex(f.expr)) = $(f_val)", x => g_val)
            ),
            time=180
        )
    end
end